tests/test_sip_proxy.py

```diff
@@ -54,47 +59,54 @@
     callback_complete = loop.create_future()
     callback_complete_proxy = loop.create_future()
 
-    async def subscribe(dialog, request):
-        assert len(dialog.peer.subscriber) == 1
-        await dialog.reply(request, status_code=200)
+    async def subscribe(request, message):
+        dialog = await request.prepare(status_code=200)
         await asyncio.sleep(0.2)
         await dialog.notify(payload='1')
 
-    async def notify(dialog, request):
-        await dialog.reply(request, 200)
-        callback_complete.set_result(request)
+    async def proxy_subscribe(request, message):
+        dialog = request._create_dialog()
+        peer = await aiosip.utils.get_proxy_peer(dialog, message)
+
+        async for proxy_response in peer.proxy_request(dialog, message, 0.1):
+            if proxy_response:
+                dialog.peer.proxy_response(proxy_response)
 
-    async def proxy_notify(dialog, request):
-        await dialog.router.proxy(dialog, request, timeout=0.1)
-        callback_complete_proxy.set_result(request)
+        # TODO: refactor
+        subscription = peer._dialogs[message.headers['Call-ID']]
 
-    app = aiosip.Application(loop=loop)
+        async for msg in subscription:
+            async for proxy_response in dialog.peer.proxy_request(subscription, msg):
+                if proxy_response:
+                    peer.proxy_response(proxy_response)
+            callback_complete_proxy.set_result(msg)
```

Comment:
@ovv @GMLudo can you guys at least go over this part, as I think this stuff impacts you guys directly?

Reply:
And yes, admittedly its a bit akward, especially since the underlying dialog was hidden, but I really think it can be cleaned up with some new abstractions targeting proxying, and I think the result here is clearer.